### DISCUSSION TITLE
Maintain and improve accessibility for screen readers

### DISCUSSION THREAD

**Discussion by papjul** [breezy-weather/breezy-weather#973] [Ideas]
If you are impacted by accessibility issues specific to screen readers, please get in touch!

You can [find existing ongoing issues here](https://github.com/breezy-weather/breezy-weather/issues?q=is%3Aissue%20state%3Aopen%20label%3AAccessibility).

**Comment by heull001** [comment]
Here are some hints what should be improved for screenreader users.

To understand all these hints, you should basically understand how screenreader users interact with the smartphone. The most used method to get information is to swipe right or left to jump to next or previous element, the new element is automaticly spoken by speech synthesis. Other methods are gestues to jump to the next headline, button or link, this element is spoken too.

I am not familar with coding android apps, so I can't give hints for implementation.

I am using this app since a few days now, so possibly there will be more suggestions in future.

If you have questions, please ask, I am willing to help if possible.

# Main screen 

## Header

The first element contains Temperature and clouds,this is good. The next three elements are graphics labeled with "feels like", "wind" and "humidity". The next three elements contain the current values of them. The next three elements are texts like the labels of the graphics before: "Feels like", "wind" and "humidity". These nine elements should be only three, one for every value,  and spoken as "<label>: <value>", for example "Feels like: 33 °F".

## Cards

The cards open when swiping on it, so there is no ability to swipe fast to the next card. As some like hourly forecast contain very much entries, it is no realistic possibility to navigate like this. Screenreader users can try to scroll on screen and hope to find the right arey, but this is only a workaround. It would be helpfull when cards open only when clicking them. If you don't want this generally, you could add an option to choose this.

The buttons for example to choose temperature or air quality for hourly forecast are identified as text, so screenreader users have to guess they are clickable. Additionally these are many buttons, so they have to switch many times right to reach the content. A dropdown list would be much better here.

# Settings

For screenreader users it is not realistic to change the order of elements by drag and drop. The best solution would be a menu with entries like "position 1: temperature" (where temperature is the element current on position 1, these and all following would be moved one position up when choosing this.)

**Reply by papjul** [reply]
Thank you very much for the detailed report!
I will review each point and get back to you if I need clarification ;)

**Reply by papjul** [reply]
I created the following issues to track the progress of each item:
- #1641 
- #1642 
- #1643 
- #1644 

Regarding drag & drop, we have plans to rewrite this screen as part of #1355, so it will be taken care of at that moment. We will also add a text in the "reorder details in header screen" to explain why some items are not be displayed despite being selected in settings.

**Comment by heull001** [comment]
As I saw in settings right now, the UV index in header is activated. It is completely ignored by the screenreader.

**Reply by papjul** [reply]
Is it night in the location you're checking? There is no UV index at night.
Because we can only display up to 4 items in header (5 in landscape), if a data is not relevant (or not provided by the source), we skip it and show the next configured one.

**Reply by heull001** [reply]
No, it's afternoon and the sun is shining. But it is not visible, I had this checked by a not visual impaired person, so this does not seem to be a accessibility problem.

**Reply by papjul** [reply]
What's the forecast source (or source for current weather if you defined one) for your location? Maybe the source doesn't provide this data. In that case, the UV index will show in the header for locations with a forecast source providing the UV index info, and it will not be shown in locations with a forecast source not providing this info.

We have a very lengthy [document describing all available data source by source](https://github.com/breezy-weather/breezy-weather/blob/main/docs/SOURCES.md). You need to expand the "Details of available data from *name of the source*" section and there will be a summary table with all available details.

Edit: Now that I think of it, the document uses a checkbox emoji to describe whether the feature is available or not, so you will probably not be able to understand if it is available or not. Let me know? We can improve accessibility of this document.

**Reply by heull001** [reply]
This was no problem, most emojis are correctly identified by screenreaders. And yes, UV index is not shown, because the national weather source doesn't support this.

**Comment by papjul** [comment]
Accessibility was improved in v5.4.3.

Here is the changelog for accessibility points:
- Main screen: details in header should now be read in the correct order by screen readers
- Main screen: fix, clicking on a day or an hour will now correctly open details on screen readers
- Main screen: improved content description of detail items, daily and hourly trend items
- Main screen: removed unused ability to tap on cards which was confusing for screen readers
- Main screen: add detailed content description of the nowcasting chart
- Main screen: make daily/hourly tabs appear as tabs in screen readers
- Main screen: added headlines to allow for quick jump between cards, on Android >= 9
- Daily details screen: improved content descriptions
- Alerts: improved content description (in Main screen and Detailed alerts screen)
- Temperature units will now be read as "degree celsius" instead of "degree C" on Android >= 7

**Reply by heull001** [reply]
This is a great improvement, thank you!

Most changes works like intended, but there are some problems still.

- The headlines on mainscreen are not spoken as headline. So screenreader users are able to jump to the next headline, but users who don't know the app would not try this, because they wouldn't know there are headlines.
- When the focus is on a card, there is no ability to jump to the next or previous headline. As the cards open by swiping on it and there is no way to close it but swiping forward or backword, focusing all elements on the card, this is very anoying, so screenreader users for example are not able to check the weather for the next day and then jump quickly to another card. Here they have to swipe back or forward or to scroll like explained in my first answer.

**Comment by heull001** [comment]
In the location list, screenreader users are not able to delete an entry. Accessibility actions (<https://developer.android.com/guide/topics/ui/accessibility/principles?hl=en#accessibility-actions>) are the best way to fix this. They would also be a good solution for changing the order of the locations and of the elements in settings. If you add accessibility actions, the hints how to delete entries and the Buttons for drag and drop should be skipped by screenreader, they could be confusing.

**Comment by heull001** [comment]
Any progress here?

**Reply by papjul** [reply]
- For the location list, there is a significant work behind this, as I mentioned previously, it will be done as part of a technical redesign in #1355 which is not a priority at the moment. I'm also not familiar with accessibility actions, so it will take additional time to learn about it first.
- For the headlines on mainscreen not spoken as headline, I'm not really sure what I should do, as I just use the native `isAccessibilityHeading` property of Android to make them appear as headlines.
- As for the inability to go to the next headline when inside a card, I didn’t have time to do some tests to better understand the issue (I need to learn the gestures to jump between headlines), but the next major version v6.0.0 will restrict the number of hourly items to a maximum of 24 hours, so that you don't have to go through a nearly infinite list of hourly data. The following hours will be summarized in the new daily details page that is the major change of this version, and more features will come in the future versions, such as daily comparisons (like "Today, temperature is 2 degrees higher than yesterday"), which should hopefully help you understand more easily the data, without having to loop hours one by one and try to remember them.

Let me know if I missed any other point?

A good thing would be to let me know for each issue your priority, so that I know what impacts you the most and focus on them first.

**Comment by RastislavKish** [comment]
Hello,

first of all, I would like to thank you for all the awesome a11y work. I have checked out the interface thoroughly and given how graphical Breezy Weather is, this is highly impressive!

Few suggestions and questions from a blind user:

* How are sighted users reviewing hourly forecast for a particular day? I can see the hourly card on the main screen, but if I understand it right, it only displays the forecast for the upcoming 24 hours, and is not connected in any way to the daily card. When I tap on a specific day, another screen opens, which according to LLM screen description features a graph, but it doesn't seem to be exposed to the screenreader.
* Accessibility actions would be highly useful for switching trends in the daily and hourly forecast. You can think of these as shortcuts to relevant actions of a particular item on the screen. For example, when you're looking at a particular day on the forecast stripe, you can at any time instantly tap on a different trend in the header to view the relevant data for that day. Blind users can't tap any element on the screen instantly, but when a component has accessibility actions, they can swipe up / down to select one and activate it by double-tapping, which is very fast, convenient, and most importantly, keeps the focus position as far as the app doesn't throw it elsewhere.
* accessibilityHeading is the correct place to mark headings, and TB should read them as so. It's weird it does not in this case, but the navigation works, and in addition, TB can also operate the UI as a table, so it's not the only way of jumping between sections.
* The labels for daily and hourly forecast could see some improvement on swiftness. I will see if I can compile the app from source and experiment with some variants to give specific suggestions.
* This is probably not a srictly a11y question, but have there been any thoughts on letting users define custom trends? I.E. displaying the data the user finds most relevant, with the ability to switch to different trends for exploring something further.

Thank you again for the great work!

**Reply by papjul** [reply]
> How are sighted users reviewing hourly forecast for a particular day? I can see the hourly card on the main screen, but if I understand it right, it only displays the forecast for the upcoming 24 hours, and is not connected in any way to the daily card. When I tap on a specific day, another screen opens, which according to LLM screen description features a graph, but it doesn't seem to be exposed to the screenreader.

The graph shows a 24-hour line chart fitting in the window width. Users who want to get details for a particular hour need to tap on the graph, and by sliding from left to right (from midnight to midnight next day), a marker shows the details above the chart as text (replacing the day/night info). When the finger is released, the day/night info is back. It's also possible to get an estimate of the value by just reading the line chart and getting an overall idea of the trends of the day.

I thought about hour details more than 24 hours later not being available to blind users, but I ran into multiple questioning/issues:
- First of all, most apps only feature the next 24 hours. Breezy Weather making the following hours available is like the cherry on the cake, as the more you advance in the future the less reliable are these hourly data. So, is it worth it to complexify the navigation on screen readers for this? Someone mentioned previously that when we were showing all hours on the main screen, it was very hard to navigate.
- The library we use for the chart shown in the daily screens doesn't have accessibility features at the moment, but it's planned. In the meantime, I didn't find a nice way to navigate the data through a screen reader (and didn't make it my priority given the previous point).

> Accessibility actions would be highly useful for switching trends in the daily and hourly forecast. You can think of these as shortcuts to relevant actions of a particular item on the screen. For example, when you're looking at a particular day on the forecast stripe, you can at any time instantly tap on a different trend in the header to view the relevant data for that day. Blind users can't tap any element on the screen instantly, but when a component has accessibility actions, they can swipe up / down to select one and activate it by double-tapping, which is very fast, convenient, and most importantly, keeps the focus position as far as the app doesn't throw it elsewhere.

The way you navigate between trends has been improved in version 6, as they should now be properly recognized as buttons and let you know which one is selected. I will review how we can integrate accessibility actions so you can read the temperature for 9 AM then switch immediately to the precipitation screen and have the screenreader read you the precipitation info for 9 AM (if I understood correctly)

> The labels for daily and hourly forecast could see some improvement on swiftness. I will see if I can compile the app from source and experiment with some variants to give specific suggestions.

Which labels are you talking about? The heading "Daily forecast"?

> This is probably not a srictly a11y question, but have there been any thoughts on letting users define custom trends? I.E. displaying the data the user finds most relevant, with the ability to switch to different trends for exploring something further.

If I understood correctly, you can add/remove/reorder trends from settings > main screen. This requires drag & drop.

**Reply by RastislavKish** [reply]
@papjul thank you for the detailed answer, this makes things much clearer!

> First of all, most apps only feature the next 24 hours. Breezy Weather making the following hours available is like the cherry on the cake, as the more you advance in the future the less reliable are these hourly data. So, is it worth it to complexify the navigation on screen readers for this?

I have no doubts you're a better expert on weather apps and their UI design than me, but honestly, I was pretty surprised and at first confused when I couldn't find hourly forecast for a particular day in the app. Aside from BW, I use Foreca, Meteoblue and Yr, all of them offering sub24 hour forecast granularity. Doesn't need to be exactly one hour, Meteoblue has three, Foreca 1, and Yr lets you choose. But it's there, so I considered this to be a standard  , and given BWs excellent reputation, I figured the mistake must be on my side. :)

Personally, I use the sub24 hour forecast a lot to better understand what is the model telling to me. It's one thing that it's going to be raining on 5th June, but I would like some more details:

* How intensively is it going to rain?
* Is it going to be a one intensive rainfall, or small showers scattered across the entire day?
* If there's a rain intensity peak, when is it going to happen? At the morning? at lunch time, at the afternoon or in the evening?
* What is the UV index and wind strenght going to be during the day? When can I expect the sunshine intensity to drop enough?

And so on. This information helps me visualise the climatic conditions and understand what the model thinks will be going on over the day, and plan accordingly. Hence for me, it's a crucial feature of any weather app.

But it's true I don't have numbers on how accurate the predictions are in practice, even the daily forecast often shifts in one direction or another. In addition, I imagine this is also location dependent and also the predictability of the expected phenomenons matters, Meteoblue even displays the weather predictability for each day. So it's certainly a good question how far ahead it actually makes sense to check these detailed forecasts.

> Someone mentioned previously that when we were showing all hours on the main screen, it was very hard to navigate.

This depends a lot on the specific design and user skill. The user complained it takes long to cross the list with Left/Right swiping. I'm not sure what the UI was at the time of writing that comment, however with the current UI, Left/Right swiping is definitely not the navigation method I'd use for moving between the cards. I either focus the card I want to use with a touch, because I know the app's layout thanks to the great description in the readme, or I use navigation by headings, that works great but TalkBack puts that mode horribly out of hand, so just pointing the screenreader to the area I want to work with is much faster.

In other words, I don't think displaying a lot of things is necessarily a problem, it's more about how you display them and whether there's a reasonable method for quick navigation. Which there is, currently.

> The library we use for the chart shown in the daily screens doesn't have accessibility features at the moment, but it's planned.

What does this mean? Are the upstream devs planning to add some a11y features? Is there any ongoing discussion about this?

> In the meantime, I didn't find a nice way to navigate the data through a screen reader

Well, one thing you could do and should be relatively easy is exposing the graph as an object with a description explaining it's a hourly graph. It would not solve the problem directly, but have two strong benefits:

* People would know the values are there, even if not accessible, and could avoid looking for them elsewhere in the app
* And TB has a gesture to send a component of the screen for analysis to Gemini, so they could get at least an idea of the graph shape, what could be a good start given the forecast reliability questions you mentioned

It's difficult for me to suggest an  actual solution without getting into technical details. For example, what you do with the daily and hourly forecasts in the main screen is absolutely perfect in my opinion, and those are also graphs. Another approach would be adding a hour slider beneath the graph, perhaps only visible by TB, and make the data area a live region, so the screenreader announces updates when the slider is moved. Indeed this would not solve the difference between holding the graph and the normal state, the hour by hour mode would either need to be active by default or activatable through an a11y action. Yr does have some graphics on the main screen you can move with a finger to slide through individual hours, and I'd swear there was a time this worked through the TalkBack's fast scroll gestures, I could just make a Right-Left or Left-Right swipe gesture on the forecast description and it would switch hour by hour. But it either doesn't work this way anymore, or I might be misremembering it, because now it doesn't work this way, I need to scroll manually, which is a bit tedious because the data are not announced in the process.

Or, a different solution could be to add an invisible list with the data to the space of the graph, this *could* be relatively simple, clean, intuitive and you could implement a11y actions on top, as I mentioned before, the main screen solution is ideal IMO.

> The way you navigate between trends has been improved in version 6, as they should now be properly recognized as buttons and let you know which one is selected.

They are, and it's cool! But I need to drift the focus in there, what's doable, but actions would make this process a lot more convenient. :)

> if I understood correctly

Yes! :)

> Which labels are you talking about? The heading "Daily forecast"?

The labels for individual days / hours. Right now, they go like:

```
Conditions, Friday, 5. June: ...
```

There are two commas and one period, which means three breaks I think we could avoid. Also, do we really need to announce the type of displayed data for each item, when it's obvious from the data itself and in addition the header shows what's currently selected? These are just some of my thoughts.

> If I understood correctly, you can add/remove/reorder trends from settings > main screen.

No. I meant something similar to what Foreca allows on its main screen, where you can exactly specify what meteo parameters are displayed for:

* Each day of the forecast
* Each hour when a daily forecast is expanded
* And the top panel displaying the current parameters.

Indeed Foreca is conceptually very different from Breezy Weather, but since some trends like conditions are already displaying several meteo parameters at once in the BW daily forecast graph, I was only curious whether there were any thoughts on letting people adjust the displayed parameters of a single trend, i.e. that I could, under daily conditions, have displayed the weather, top temperature, feels like temperature, min temperature, amount of rainfall and UV index. Those are 6 parameters, the same number the conditions trend displays ATM, but is customized to individual needs, and if I wanted to check other parameters, I could just switch the trends using the card's header bar.

I apologize if the question does not conceptually make sense, I have little information on how exactly is BW plotting multi-parameter charts, so I have no estimates if this is a real option or not. :)

Thank you for all the amazing work!

**Reply by papjul** [reply]
> @papjul thank you for the detailed answer, this makes things much clearer!
> 
> > First of all, most apps only feature the next 24 hours. Breezy Weather making the following hours available is like the cherry on the cake, as the more you advance in the future the less reliable are these hourly data. So, is it worth it to complexify the navigation on screen readers for this?
> 
> I have no doubts you're a better expert on weather apps and their UI design than me, but honestly, I was pretty surprised and at first confused when I couldn't find hourly forecast for a particular day in the app. Aside from BW, I use Foreca, Meteoblue and Yr, all of them offering sub24 hour forecast granularity. Doesn't need to be exactly one hour, Meteoblue has three, Foreca 1, and Yr lets you choose. But it's there, so I considered this to be a standard , and given BWs excellent reputation, I figured the mistake must be on my side. :)

It's not just an UI thing; sources rarely provide the hourly data 24 hours ahead, or limited hourly (for example, 3 hours by 3 hours for the next 48 hours, 6 hours by 6 hours for the next 72 hours). Foreca and Meteoblue are premium sources we don't support because they cost 1000 dollars a year. Yr is based on Met Norway which we support and is one of the rare sources with hourly data for 10 days. Another one is Open-Meteo for 16 days.

As for the navigation, the issue is more that you would have to be patient to have the screenreader read you every single hour of the 24 hours chart of that day. The chart is just one big block, compared to the main screen where the navigation can be made 1 hour by 1 hour.

> > The library we use for the chart shown in the daily screens doesn't have accessibility features at the moment, but it's planned.
> 
> What does this mean? Are the upstream devs planning to add some a11y features? Is there any ongoing discussion about this?

There is a very nice basic accessibility support merged in patrykandpatrick/vico#1069, but it is not currently released, because only Android Compose is supported and they want feature parity with Android views (two different technologies for displaying charts). Breezy Weather mainly uses Compose. It does use views in the main screen in the nowcasting block, but that block already has a content description which is a summary of when the rain is starting and when it is ending, so having accessibility support for Compose would cover all cases for us if it was released.

> They are, and it's cool! But I need to drift the focus in there, what's doable, but actions would make this process a lot more convenient. :)

Tracked as #2798 

> > In the meantime, I didn't find a nice way to navigate the data through a screen reader
> 
> Well, one thing you could do and should be relatively easy is exposing the graph as an object with a description explaining it's a hourly graph. It would not solve the problem directly, but have two strong benefits:
> 
>     * People would know the values are there, even if not accessible, and could avoid looking for them elsewhere in the app
> 
>     * And TB has a gesture to send a component of the screen for analysis to Gemini, so they could get at least an idea of the graph shape, what could be a good start given the forecast reliability questions you mentioned

Agreed. Tracked as #2799 

> > Which labels are you talking about? The heading "Daily forecast"?
> 
> The labels for individual days / hours. Right now, they go like:
> 
> ```
> Conditions, Friday, 5. June: ...
> ```
> 
> There are two commas and one period, which means three breaks I think we could avoid. Also, do we really need to announce the type of displayed data for each item, when it's obvious from the data itself and in addition the header shows what's currently selected? These are just some of my thoughts.

There's a comma between Conditions and Friday, because the syntax would not be correct (at least my very basic screen reader reads the two words as if there are in the same sentence and it goes from Conditions to Friday with no pause, and it feels just weird). I could replace it by a colon separator, if that would be better?

As for the date, the syntax comes from your language. We don't really have much choice there, it's an internationalization library that is taking care of spelling the dates in the correct order and with the week days and months translated. In French, it would be read "Friday 5 June" with no comma or dot.

> > If I understood correctly, you can add/remove/reorder trends from settings > main screen.
> 
> No. I meant something similar to what Foreca allows on its main screen, where you can exactly specify what meteo parameters are displayed for:
> 
>     * Each day of the forecast
> 
>     * Each hour when a daily forecast is expanded
> 
>     * And the top panel displaying the current parameters.
> 
> 
> Indeed Foreca is conceptually very different from Breezy Weather, but since some trends like conditions are already displaying several meteo parameters at once in the BW daily forecast graph, I was only curious whether there were any thoughts on letting people adjust the displayed parameters of a single trend, i.e. that I could, under daily conditions, have displayed the weather, top temperature, feels like temperature, min temperature, amount of rainfall and UV index. Those are 6 parameters, the same number the conditions trend displays ATM, but is customized to individual needs, and if I wanted to check other parameters, I could just switch the trends using the card's header bar.
> 
> I apologize if the question does not conceptually make sense, I have little information on how exactly is BW plotting multi-parameter charts, so I have no estimates if this is a real option or not. :)

I went on the Foreca website, and if I understood correctly: no, it's not possible to display multiple parameters at once in the charts. Both from the technical conception and from the visual perspective.

Honestly, the Foreca chart feels overwhelming and not easy to read with all these lines and data from all sides. I can understand that with some experience, one could probably be more efficient by having all the data at once than with different charts, but one chart for each kind of data is probably more accessible for most people. Also, our daily charts are split in two (day and night), which mean there are already twice the data most weather apps usually show, so adding more would make the chart very cluttered.

There are plans to make an hourly table of all data to allow comparison of two locations side-by-side in the long term, that could be a workaround.

**Comment by RastislavKish** [comment]
One more thing, @heull001 drag and drop is problematic on keyboard controlled systems, but on a touch screen you can do this pretty easily even with a screenreader. Just find the item you want to move using explore by touch, then tap and hold, move the finger to the position where you want it and lift the finger. Note you don't necessarily need to hit an exact position in the list, it's often enough to just move an item a position up or down, also instead of trying to move the item you want to a specific location, you can instead just move the things you don't want out of the way, if you fear you'd break the layout, and then just fine-tune one-place swaps. I have reordered the lists several times according to my current preference several times and it has always worked reliably.

I don't think the reordering is going to require any special UI for screenreader users. If you want to make things extra convenient, you can introduce accessibility actions for moving the focused item one position up / down and to the beginning or end of the list. But this is more a luxury thing IMO, unlike the daily and hourly forecast cards, where the actions would have strong benefits.
